src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -0,0 +1,146 @@
+           "store more.";
+
+    num_bytes_task_events_ -= task_events_[next_idx_to_overwrite_].ByteSizeLong();
+    num_bytes_task_events_ += events_by_task.ByteSizeLong();
+
+    // Change the underlying storage.
+    auto &to_replaced = task_events_.at(next_idx_to_overwrite_);
+    std::swap(to_replaced, events_by_task);
+    auto replaced = events_by_task;
+
+    // Update index.
+    TaskAttempt replaced_attempt = std::make_pair<>(
+        TaskID::FromBinary(replaced.task_id()), replaced.attempt_number());
+    task_attempt_index_.erase(replaced_attempt);
+    task_attempt_index_[task_attempt] = next_idx_to_overwrite_;
+
+    // Update iter.
+    next_idx_to_overwrite_ = (next_idx_to_overwrite_ + 1) % max_num_task_events_;
+
+    return replaced;
+  }
+
+  // Add to index.
+  task_attempt_index_[task_attempt] = task_events_.size();
+  // Add a new task events.
+  task_events_.push_back(std::move(events_by_task));
+  return absl::nullopt;
+}
+
+void GcsTaskManager::HandleAddTaskEventData(rpc::AddTaskEventDataRequest request,
+                                            rpc::AddTaskEventDataReply *reply,
+                                            rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(DEBUG) << "Adding task state event:" << request.data().ShortDebugString();
+  // Dispatch to the handler
+  io_service_.post(
+      [this, reply, send_reply_callback, request = std::move(request)]() {
+        auto data = request.data();
+        size_t num_to_process = data.events_by_task_size();
+        // Update counters.
+        total_num_profile_task_events_dropped_ += data.num_profile_task_events_dropped();
+        total_num_status_task_events_dropped_ += data.num_status_task_events_dropped();
+
+        for (auto &events_by_task : *data.mutable_events_by_task()) {
+          total_num_task_events_reported_++;
+          auto task_id = TaskID::FromBinary(events_by_task.task_id());
+          // TODO(rickyx): add logic to handle too many profile events for a single task
+          // attempt.
+          auto replaced_task_events =
+              task_event_storage_->AddOrReplaceTaskEvent(std::move(events_by_task));
+
+          if (replaced_task_events) {
+            if (replaced_task_events->has_state_updates()) {
+              // TODO(rickyx): should we un-flatten the status updates into a list of
+              // StatusEvents? so that we could get an accurate number of status change
+              // events being dropped like profile events.
+              total_num_status_task_events_dropped_++;
+            }
+            if (replaced_task_events->has_profile_events()) {
+              total_num_profile_task_events_dropped_ +=
+                  replaced_task_events->profile_events().events_size();
+            }
+          }
+          RAY_LOG(DEBUG) << "Processed a task event. [task_id=" << task_id.Hex() << "]";
+        }
+        RAY_LOG_EVERY_MS(INFO, 10000)
```

Comment:
maybe just adding this into debug_state_gcs.txt instead of periodically printing it?